packages/superset-ui-chart/src/components/ChartDataProvider.tsx

```diff
@@ -0,0 +1,147 @@
+/* eslint react/sort-comp: 'off' */
+import React, { ReactNode } from 'react';
+import { SupersetClientInterface, RequestConfig } from '../../../superset-ui-connection/src/types';
+
+import ChartClient, { SliceIdAndOrFormData } from '../clients/ChartClient';
+import { ChartFormData } from '../types/ChartFormData';
+import { QueryData } from '../models/ChartProps';
+
+type Datasource = object;
+
+interface Payload {
+  formData: Partial<ChartFormData>;
+  queryData: QueryData;
+  datasource?: Datasource;
+}
+
+export interface ProvidedProps {
+  payload?: Payload;
+  error?: Error;
+  loading?: boolean;
+}
+
+export type Props =
+  /** User can pass either one or both of sliceId or formData */
+  SliceIdAndOrFormData & {
+    /** Child function called with ProvidedProps */
+    children: (provided: ProvidedProps) => ReactNode;
+    /** Superset client which is used to fetch data. It should already be configured and initialized. */
+    client?: SupersetClientInterface;
+    /** Will fetch and include datasource metadata for SliceIdAndOrFormData in the payload. */
+    loadDatasource?: boolean;
+    /** Callback when an error occurs. Enables wrapping the Provider in an ErrorBoundary. */
+    onError?: (error: ProvidedProps['error']) => void;
+    /** Callback when data is loaded. */
+    onLoaded?: (payload: ProvidedProps['payload']) => void;
+    /** Hook to override the formData request config. */
+    formDataRequestOptions?: Partial<RequestConfig>;
+    /** Hook to override the datasource request config. */
+    datasourceRequestOptions?: Partial<RequestConfig>;
+    /** Hook to override the queryData request config. */
+    queryRequestOptions?: Partial<RequestConfig>;
+  };
+
+type State = {
+  status: 'unitialized' | 'fetching' | 'error' | 'loaded';
```

Comment:
typo `unitialized` => `uninitialized`

Reply:
Should we defined enum for `status`?

Reply:
Another minor: `loading` & `loaded` vs. `fetching` & `loaded`?

Reply:
thanks for catching the typos. `loading` + `loaded` seem good instead of the `fetching` inconsistency.

I thought generally we were avoiding enums? (esp here seems overkill vs simple union since it's scoped to the component)

Reply:
sure. no strong opinion on `enum`. strings with type checked seems good enough for me.